libglusterfs/src/monitoring.c

```diff
@@ -90,46 +92,46 @@ dump_latency_and_count(xlator_t *xl, int fd)
     if ((xl != xl->ctx->primary) && (xl->ctx->active != xl->graph))
         return;
 
-    count = GF_ATOMIC_GET(xl->stats.total.count);
-    dprintf(fd, "%s.total.fop-count %" PRIu64 "\n", xl->name, count);
-
-    count = GF_ATOMIC_GET(xl->stats.interval.count);
-    dprintf(fd, "%s.interval.fop-count %" PRIu64 "\n", xl->name, count);
-    GF_ATOMIC_INIT(xl->stats.interval.count, 0);
-
     for (index = 0; index < GF_FOP_MAXVALUE; index++) {
-        fop = GF_ATOMIC_GET(xl->stats.total.metrics[index].fop);
+        fop = GF_ATOMIC_GET(xl->stats[index].total_fop);
         if (fop) {
             dprintf(fd, "%s.total.%s.count %" PRIu64 "\n", xl->name,
                     gf_fop_list[index], fop);
+            total_fop_count += fop;
         }
-        fop = GF_ATOMIC_GET(xl->stats.interval.metrics[index].fop);
+        GF_ATOMIC_SWAP(xl->stats[index].interval_fop, fop);
         if (fop) {
             dprintf(fd, "%s.interval.%s.count %" PRIu64 "\n", xl->name,
                     gf_fop_list[index], fop);
+            interval_fop_count += fop;
+            fop = 0;
         }
-        cbk = GF_ATOMIC_GET(xl->stats.interval.metrics[index].cbk);
+        GF_ATOMIC_SWAP(xl->stats[index].interval_fop_cbk, cbk);
         if (cbk) {
             dprintf(fd, "%s.interval.%s.fail_count %" PRIu64 "\n", xl->name,
                     gf_fop_list[index], cbk);
+            total_fop_cbk += cbk;
+            cbk = 0;
         }
-        if (xl->stats.interval.latencies[index].count != 0) {
+        if (xl->stats[index].latencies.count != 0) {
             dprintf(fd, "%s.interval.%s.latency %lf\n", xl->name,
                     gf_fop_list[index],
-                    (((double)xl->stats.interval.latencies[index].total) /
-                     xl->stats.interval.latencies[index].count));
+                    (((double)xl->stats[index].latencies.total) /
+                     xl->stats[index].latencies.count));
             dprintf(fd, "%s.interval.%s.max %" PRIu64 "\n", xl->name,
-                    gf_fop_list[index],
-                    xl->stats.interval.latencies[index].max);
+                    gf_fop_list[index], xl->stats[index].latencies.max);
             dprintf(fd, "%s.interval.%s.min %" PRIu64 "\n", xl->name,
-                    gf_fop_list[index],
-                    xl->stats.interval.latencies[index].min);
+                    gf_fop_list[index], xl->stats[index].latencies.min);
         }
-        GF_ATOMIC_INIT(xl->stats.interval.metrics[index].cbk, 0);
-        GF_ATOMIC_INIT(xl->stats.interval.metrics[index].fop, 0);
+        GF_ATOMIC_INIT(xl->stats[index].interval_fop_cbk, 0);
+        GF_ATOMIC_INIT(xl->stats[index].interval_fop, 0);
+        memset(&xl->stats[index].latencies, 0,
```

Comment:
This is not safe, but it's not easy to change and it was already done this way before this patch, so I guess we can continue with this as it is for now, until we find a better way.